src/app/ReadClient.cpp

```diff
@@ -894,32 +913,49 @@ void ReadClient::OnLivenessTimeoutCallback(System::Layer * apSystemLayer, void *
                  "Subscription Liveness timeout with SubscriptionID = 0x%08" PRIx32 ", Peer = %02x:" ChipLogFormatX64,
                  _this->mSubscriptionId, _this->GetFabricIndex(), ChipLogValueX64(_this->GetPeerNodeId()));
 
+    if (_this->mIsPeerICD)
+    {
+        // If the device is idle, we mark the subscription as "InactiveICDSubscription", readClient consumer decides whether it goes
+        // with retry or resubscription is triggered on `OnActiveModeNotification`. Note: the liveness timeout is always longer than
+        // the MaxInterval (and idle duration), so we can move the device to `IdleSubcription` when liveness timeout is reached.
+        ChipLogProgress(DataManagement, "Peer is not active now, mark the subscription as InactiveICDSubscription.");
+        _this->MoveToState(ClientState::InactiveICDSubscription);
+        _this->Close(CHIP_ERROR_ICD_SUBSCRIBE_INACTIVE_TIMEOUT);
+        // Return early here since the peer is idle and cannot receive any messages. The resubscription will be deferred until
+        // `OnActiveModeNotification` is invoked.
+        return;
+    }
+
+    _this->TriggerResubscriptionForLivenessTimeout();
+}
+
+void ReadClient::TriggerResubscriptionForLivenessTimeout()
+{
     // We didn't get a message from the server on time; it's possible that it no
     // longer has a useful CASE session to us.  Mark defunct all sessions that
     // have not seen peer activity in at least as long as our session.
-    const auto & holder = _this->mReadPrepareParams.mSessionHolder;
+    const auto & holder = mReadPrepareParams.mSessionHolder;
     if (holder)
     {
         System::Clock::Timestamp lastPeerActivity = holder->AsSecureSession()->GetLastPeerActivityTime();
-        _this->mpImEngine->GetExchangeManager()->GetSessionManager()->ForEachMatchingSession(
-            _this->mPeer, [&lastPeerActivity](auto * session) {
-                if (!session->IsCASESession())
-                {
-                    return;
-                }
+        mpImEngine->GetExchangeManager()->GetSessionManager()->ForEachMatchingSession(mPeer, [&lastPeerActivity](auto * session) {
+            if (!session->IsCASESession())
+            {
+                return;
+            }
 
-                if (session->GetLastPeerActivityTime() > lastPeerActivity)
-                {
-                    return;
-                }
+            if (session->GetLastPeerActivityTime() > lastPeerActivity)
+            {
+                return;
+            }
 
-                session->MarkAsDefunct();
-            });
+            session->MarkAsDefunct();
```

Comment:
@bzbarsky-apple this seems to be existing code. So I have not yet thought it well on this part. I would leave it a follow-up here. thanks

Reply:
https://github.com/project-chip/connectedhomeip/issues/30903

Reply:
It's not "existing code".  It's the new code _not matching the existing code_, no?

As far as I am concerned, this needs to be addressed before this can land, since it seems like a fundamental architectural problem with the approach.

Reply:
resolved. thanks